examples/linux/io_epoll_test.cpp

```diff
@@ -181,49 +160,50 @@ int main() {
   auto buffer = std::vector<char>{};
   buffer.resize(1);
   try {
-    sync_wait(
-      when_all(
-        // write chunk of data into one end repeatedly
-        pipe_write(wPipe, databuffer, scheduler, stopWrite.get_token()),
-        // read the data 1 byte at a time from the other end
-        sequence(
-          // read for some time before starting measurement
-          // this is done to reduce startup effects
-          pipe_bench(rPipe, buffer, scheduler, WARMUP_DURATION, data, reps, offset),
-          // reset measurements to exclude warmup
-          lazy([&]{
-            // restart reps and keep offset in data
-            offset = reps%sizeof(data);
-            reps = 0;
-            printf("warmup completed!\n");
-            // exclude the warmup time
-            start = end = std::chrono::high_resolution_clock::now();
-          }),
-          // do more reads and measure how many reads occur
-          pipe_bench(rPipe, buffer, scheduler, BENCHMARK_DURATION, data, reps, offset),
-          // report results
-          lazy([&]{
-            end = std::chrono::high_resolution_clock::now();
-            printf("benchmark completed!\n");
-            auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
-                    end - start)
-                    .count();
-            auto ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
-                    end - start)
-                    .count();
-            double reads = 1000000000.0 * reps / ns;
-            std::cout
-                << "completed in "
-                << ms << " ms, "
-                << ns << "ns, "
-                << reps << "ops\n";
-            std::cout
-                << "stats - "
-                << reads << "reads, "
-                << ns/reps << "ns-per-op, "
-                << reps/ms << "ops-per-ms\n";
-            stopWrite.request_stop();
-          }))));
+    when_all(
+      // write chunk of data into one end repeatedly
+      pipe_write(wPipe, databuffer, scheduler, stopWrite.get_token()),
+      // read the data 1 byte at a time from the other end
+      sequence(
+        // read for some time before starting measurement
+        // this is done to reduce startup effects
+        pipe_bench(rPipe, buffer, scheduler, WARMUP_DURATION, data, reps, offset),
+        // reset measurements to exclude warmup
+        [&] {
+          // restart reps and keep offset in data
+          offset = reps%sizeof(data);
+          reps = 0;
+          printf("warmup completed!\n");
+          // exclude the warmup time
+          start = end = std::chrono::high_resolution_clock::now();
```

Comment:
I wonder if the time-measurement here could be handled with some kind of reusable `print_time_taken()` algorithm that you can wrap a sender it to print the time it took to complete.

e.g. something like:
```
auto print_time_taken(sender auto s, std::string tag = {}) {
  using namespace std::chrono;
  return
    lazy([] { return steady_clock::now(); })
    | let([s=std::move(s), tag=std::move(tag)]
          (const clock_t::time_point& start) mutable {
      return std::move(s)
        | finally(lazy([&] {
            auto timeTaken = steady_clock::now() - start;
            std::stringstream ss;
            if (!tag.empty()) ss << "[" << tag << "] ";
            ss << "completed in " << duration_cast<milliseconds>(timeTaken).count() << "ms, "
               << duration_cast<nanoseconds>(timeTaken).count() << "ns\n";
            std::puts(std::move(ss).str().c_str());
          }));
    });
}
```

Reply:
I love this idea. I am happy for that to be a separate PR

Reply:
Would it make sense to factor this such that `print_time_taken` is a composition of something like `compute_time_taken` and `print`?